### PR TITLE
LibWeb+LibGfx: Implement StackBlurFilter and use for shadow painting (and some fun fixes)

### PR DESCRIPTION
**Pull request by MacDue**
| Before                                                                                                                                        | After                                                                                                                                         | Chromium                                                                                                                                      |
|-----------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|
| Screenshot from 2022-06-29 19-13-31 | Screenshot from 2022-06-29 19-11-34 | image                               |
| Screenshot from 2022-06-28 11-33-40 | Screenshot from 2022-06-28 11-32-57 | Screenshot from 2022-06-28 11-33-55 |
| image                               | image                               | N/A                                                                                                                                           |

**TLDR:  it's more good-er and more vroom vroom brrr!**

**boogs fixed:**

- 'Halo' around shaodows
- Off-by-one for shadows non integer positioned elements 
- Box shadows on elements where the width/2 or height/2 is < the blur radius
- Faster shadow painting brrrrrrrrrrrrr

### CODE REVIEW

**Comment by MacDue** [comment]
Gave up trying to make `FastBoxBlurFilter` actually fast, since the main slowness was really from it requiring 3 entire passes over the image + a lot more allocations. `StackBlurFilter` achieves a good result in just one pass and with less memory.

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+#    pragma GCC optimize("O3")
+#endif
+
+#include <AK/Array.h>
+#include <AK/Vector.h>
+#include <LibGfx/Filters/StackBlurFilter.h>
+
+namespace Gfx {
+
+constexpr Array<u16, 255> mult_table {
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:18]: Can we get an equation for these tables?

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+    Vector<Color, 512> m_data;
+};
+
+// This is an implementation of StackBlur by Mario Klingemann (https://observablehq.com/@jobleonard/mario-klingemans-stackblur)
+// (Link is to a secondary source as the original site is now down)
+FLATTEN void StackBlurFilter::process_rgba(u8 radius)
+{
+    // TODO: Implement a plain RGB version of this (if required)
+
+    using uint = unsigned;
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:119]: Can we use u32 in all cases instead?

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+    for (uint y = 0; y < height; y++) {
+        stack_iterator = stack_start;
+
+        auto color = get_pixel(0, y);
+        for (uint i = 0; i < radius_plus_1; i++)
+            *(stack_iterator++) = color;
+
+        // All the sums here work to approximate a gaussian.
+        uint red_in_sum = 0;
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:161]: Not mandatory:
These could easily use some SIMD types, to possibly paralleize the multiplies and additions
It would also condense the code/avoid repetition of nearly the same code

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        }
+
+        ALWAYS_INLINE Color* operator->()
+        {
+            return &m_data.at(m_idx);
+        }
+
+        ALWAYS_INLINE Iterator operator++()
+        {
+            // Note: This seemed to profile slightly better than %
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:79]: Note:
Modulo is a division in disguise, which usually is quite expensive

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp [hunk 5]
```diff
@@ -284,7 +284,7 @@ FLATTEN void StackBlurFilter::process_rgba(u8 radius)
             if (alpha != 0)
                 set_pixel(x, y, Color((red_sum * sum_mult) >> sum_shift, (green_sum * sum_mult) >> sum_shift, (blue_sum * sum_mult) >> sum_shift, alpha));
             else
-                set_pixel(x, y, transparent_white);
+                set_pixel(x, y, fill_color);
```
Comment (Hendiadyoin1) [inline Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:287]: wouldn't it be reasonable to just not set a color in this case?

**Review by Hendiadyoin1 — COMMENTED** [review]
Nice work overall!

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp (hunk 2 above)
Comment (MacDue) [reply Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:119]: I kinda prefer using `uint` here since `u32` (to me anyway) suggests that it needs 32 bits. It actually only needs up to 17 bits here.

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp (hunk 1 above)
Comment (MacDue) [reply Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:18]: I can probably try to derive it later (and maybe replace these with some `constexpr` functions). The tables anyway are used to approximate a division by radius^2

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp (hunk 3 above)
Comment (MacDue) [reply Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:161]: I'll give that a go later today and see how that effects the performance

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp (hunk 2 above)
Comment (Hendiadyoin1) [reply Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:119]: well thats unfortunate, just off a `short`...
lets let a mainteainer make the call here

Comment (Hendiadyoin1) [reply Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:119]: A comment about the used size would be nice, if we know that though

Comment (MacDue) [reply Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:119]: I think if you limit the max radius to 180px it fits in a `u16`. Maybe that's the thing to do since any blur above about 20px is super rare (and I'm not sure a really high radius above 180px even makes sense for a box-shadow).

Comment (alimpfard) [reply Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:119]: This just seems excessive, why not use `unsigned` instead of giving it a name that means exactly the same thing?

Comment (MacDue) [reply Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:119]: Because `unsigned` is annoyingly long (and I always type it wrong)

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp (hunk 5 above)
Comment (MacDue) [reply Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:271]: Not necessarily I think it is possible to blur to nothing with a high enough radius (and removing these makes no difference in the profile, so I see no reason to).

File: Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp (hunk 3 above)
Comment (MacDue) [reply Userland/Libraries/LibGfx/Filters/StackBlurFilter.cpp:145]: ...had enough other issues crop up that I don't want to look at SIMD now too. This yak is free for the taking.

File: AK/Math.h [hunk 6]
```diff
@@ -665,6 +665,17 @@ constexpr T pow(T x, T y)
     return exp2<T>(y * log2<T>(x));
 }
 
+template<FloatingPoint T>
+constexpr T ceil(T num)
+{
+    if (is_constant_evaluated()) {
+        return (static_cast<double>(static_cast<i64>(num)) == num)
```
[the remark this answers is not in the document]
Comment (MacDue) [reply AK/Math.h:672]:
```c++
        if (num < NumericLimits<i64>::min() || num > NumericLimits<i64>::max())
            return num;
```
Like so?

[the remark this answers is not in the document]
Comment (MacDue) [reply AK/Math.h:675]: (I think I've done this)
